### ISSUE TITLE
JavaScript grammar for JavaScript target fails on tree build

### ISSUE THREAD

**Issue by aherlihy** [antlr/grammars-v4#1580]
Hi, I'm trying to build and visit a JavaScript (ECMAScript 6+) tree in the JavaScript target, but the tree-building step hangs and eventually fails with "Maximum call stack size exceeded". I have it working for the [ECMAScript](https://github.com/antlr/grammars-v4/blob/master/ecmascript/JavaScript/ECMAScript.g4) and [Python3](https://github.com/antlr/grammars-v4/blob/master/python3-js/Python3.g4) grammars (which is why I'm not using Esprima or another JS parser since I need multi-language support).

I build via:

`java -Xmx500M -cp '/usr/local/lib/antlr-4.7.2-complete.jar:$CLASSPATH' org.antlr.v4.Tool  -Dlanguage=JavaScript -lib antlr -o antlr -no-listener -visitor -Xexact-output-dir antlr/JavaScriptLexer.g4`

then

`java -Xmx500M -cp '/usr/local/lib/antlr-4.7.2-complete.jar:$CLASSPATH' org.antlr.v4.Tool  -Dlanguage=JavaScript -lib antlr -o antlr -no-listener -visitor -Xexact-output-dir antlr/JavaScriptParser.g4`

Which seems to generate the correct files (JavaScriptParser.interp, JavaScriptLexer.interp, JavaScriptLexer.tokens, JavaScriptParser.tokens, JavaScriptParser.js, JavaScriptLexer.js, JavaScriptParserVisitor.js) into the same directory with [JavaScriptBaseParser.js](https://github.com/antlr/grammars-v4/blob/master/javascript/JavaScript/JavaScriptBaseParser.js) and [JavaScriptBaseLexer.js](https://github.com/antlr/grammars-v4/blob/master/javascript/JavaScript/JavaScriptBaseLexer.js).

I try to construct the tree via:

```js
const JavaScriptLexer = require('../antlr/JavaScriptLexer.js').JavaScriptLexer;
const JavaScriptParser = require('../antlr/JavaScriptParser.js').JavaScriptParser;
const JavaScriptVisitor = require('../antlr/JavaScriptParserVisitor.js').JavaScriptParserVisitor;

const input = "1";
const chars = new antlr4.InputStream(input);
const lexer = new JavaScriptLexer(chars);
lexer.strictMode = false;
const tokens = new antlr4.CommonTokenStream(lexer);
const parser = new JavaScriptParser(tokens);
parser.buildParseTrees = true;
const tree = parser.program();
```

which hangs until I get the error:

```
RangeError: Maximum call stack size exceeded
    at ParserATNSimulator.ruleTransition (.../node_modules/antlr4/atn/ParserATNSimulator.js:1494:55)
    at ParserATNSimulator.getEpsilonTarget (.../node_modules/antlr4/atn/ParserATNSimulator.js:1391:21)
    at ParserATNSimulator.closure_ (.../node_modules/antlr4/atn/ParserATNSimulator.js:1273:22)
    at ParserATNSimulator.closureCheckingStopState (.../node_modules/antlr4/atn/ParserATNSimulator.js:1254:10)
    at ParserATNSimulator.closure_ (.../node_modules/antlr4/atn/ParserATNSimulator.js:1310:18)
    at ParserATNSimulator.closureCheckingStopState (.../node_modules/antlr4/atn/ParserATNSimulator.js:1254:10)
    at ParserATNSimulator.closure_ (.../node_modules/antlr4/atn/ParserATNSimulator.js:1310:18)
    at ParserATNSimulator.closureCheckingStopState (.../node_modules/antlr4/atn/ParserATNSimulator.js:1254:10)
    at ParserATNSimulator.closure_ (.../node_modules/antlr4/atn/ParserATNSimulator.js:1310:18)
    at ParserATNSimulator.closureCheckingStopState (.../node_modules/antlr4/atn/ParserATNSimulator.js:1254:10)
```

So I'm not able to get a tree that I could use for the visitor. It fails for every input string, regardless of how complex. I assume there is support for the JS grammar in the JS target since the directory exists, but I'm wondering if I'm doing something wrong with how I'm setting things up.

Any help would be greatly appreciated!

**Comment by studentmain** [comment]
Run your code with debugger and/or limit NodeJS's stack size, you can get more useful stack trace.

(But if you can use babel/esprima/acorn/meriyah...... why do you still use ANTLR in JS to parse JS?)

**Comment by aherlihy** [comment]
I cannot use babel/asprima/acorn because I need multi-language support. If you'd like to see more details, https://github.com/mongodb-js/bson-transpilers.

The debugger in Webstorm provides:

```
Debugger listening on ws://127.0.0.1:52261/57e35b0c-2c2c-4168-bcb7-ec5fd422cd89
For help, see: https://nodejs.org/en/docs/inspector
Debugger attached.
.../antlr/JavaScriptParser.js:1024
	    	throw re;
	    	^

RangeError: Maximum call stack size exceeded
    at SingletonPredictionContext.PredictionContext.updateHashCode (.../node_modules/antlr4/PredictionContext.js:68:54)
    at Hash.update (.../node_modules/antlr4/Utils.js:357:27)
    at new SingletonPredictionContext (.../node_modules/antlr4/PredictionContext.js:116:8)
    at Function.SingletonPredictionContext.create (.../node_modules/antlr4/PredictionContext.js:132:10)
    at ParserATNSimulator.ruleTransition (.../node_modules/antlr4/atn/ParserATNSimulator.js:1499:49)
    at ParserATNSimulator.getEpsilonTarget (.../node_modules/antlr4/atn/ParserATNSimulator.js:1391:21)
    at ParserATNSimulator.closure_ (.../node_modules/antlr4/atn/ParserATNSimulator.js:1273:22)
    at ParserATNSimulator.closureCheckingStopState (.../node_modules/antlr4/atn/ParserATNSimulator.js:1254:10)
    at ParserATNSimulator.closure_ (.../node_modules/antlr4/atn/ParserATNSimulator.js:1310:18)
    at ParserATNSimulator.closureCheckingStopState (.../node_modules/antlr4/atn/ParserATNSimulator.js:1254:10)
Waiting for the debugger to disconnect...

Process finished with exit code 1
```

Changing the stack size using --stack-size=500 did not change anything. Is that what you're looking for or is there something else I can provide?

**Comment by studentmain** [comment]
> Is that what you're looking for

The bottom side (not top side, which is full of runtime function call) of call stack, or if you can, entire call stack. So we can see it failed at which rule. Try to use --stack-size=50 or smaller.

**Comment by aherlihy** [comment]
Could you please provide me the exact command you'd like me to run? I have provided the entire call stack, even with --stack-size 50 there is not more to the error message than what I've written. 

Alternatively I made a tiny project: https://github.com/aherlihy/test-js-antlr that shows the bug. If you clone it and try to run it then it should show you exactly what I'm seeing. Thank you!

**Comment by aherlihy** [comment]
@studentmain Just a ping, I would love an update on this if possible. Thank you!

**Comment by studentmain** [comment]
![图片](https://user-images.githubusercontent.com/17123090/74669910-03e24880-51a0-11ea-95a6-ea0ca1d08186.png)

Yes, it's strange... I think it looks like a runtime bug.

**Comment by aherlihy** [comment]
Is there anything that can be done? I don't have any problems with the ecmascript grammar, but unfortunately we need to support modern javascript. Since the JS target exists it must have worked at some point right?

**Comment by studentmain** [comment]
That would be difficult, diff and backport all modern js feature to es5 grammar might be a solution. Try typescript target might be useful too. Another possiblity is base class,  And if they don't work, don't forget the final option: write special code for every platform.

JS target (in fact, all target except Java) are not well tested because of lack of infrastructure support (unit test support, IDE support, etc). I guess last worked version is before #1553 (I haven't test JS target in that PR).

**Comment by ericvergnaud** [comment]
Fixed by #3077
